### PR TITLE
2.6 - Fix Model Cache Restart and All-Watcher Shutdown

### PR DESCRIPTION
**Pull request by manadart**
## Description of change

This patch does 2 main things:
- If a model is removed, it prevents a running all-watcher from closing.
- If the watcher backing the cache worker is shut down cleanly, that worker exists - for any other error, a new watcher is created and the worker proceeds.

## QA steps

Reproduce:
- With 2.6.3, bootstrap.
- Run `juju debug log -m controller` in a separate terminal.
- `juju destroy-model default -y` and observe the "model not found" errors.

Confirm:
- Same as above and observe that the cache does not bounce, and the "not found" errors are gone.

## Documentation changes

None.

## Bug reference

https://bugs.launchpad.net/juju/+bug/1832294

### CODE REVIEW

**Comment by manadart** [comment]
[Model Cache Worker Restart Behaviour](https://trello.com/c/Kal5vIKC/1507-model-cache-worker-restart-behaviour)

File: state/allwatcher.go [hunk 1]
```diff
@@ -1573,9 +1573,8 @@ func (b *allModelWatcherStateBacking) Changed(all *multiwatcherStore, change wat
 
 	st, err := b.getState(modelUUID)
 	if err != nil {
-		if exists, modelErr := b.st.ModelExists(modelUUID); exists && modelErr == nil {
-			// The entity's model is gone so remove the entity
-			// from the store.
+		if errors.IsNotFound(err) {
```
Comment (manadart) [inline state/allwatcher.go]: This is the important part to stop the all-watcher exiting.

A not-found error would previously bubble up stopping the watcher, which it shouldn't - the watcher now just deals with the fact that the model is gone.

File: state/multiwatcher.go [hunk 2]
```diff
@@ -318,18 +318,26 @@ func (sm *storeManager) respond() {
 		changes := sm.all.ChangesSince(revno)
 		if len(changes) == 0 {
 			if req.noChanges != nil {
-				req.noChanges <- struct{}{}
+				select {
```
Comment (manadart) [inline state/multiwatcher.go]: Changes in this method are just some drive-by defensive additions.

**Comment by howbazaar** [comment]
Extracted the allwatcher and multiwatcher fixes and added a test. Landing against 2.5 here #10321 .

**Comment by howbazaar** [comment]
Superseded by #10322
